src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
+            if (_projects.Count == 0)
             {
                 throw new GracefulException(CommonLocalizableStrings.SpecifyAtLeastOneProjectToAdd);
             }
-
-            PathUtility.EnsureAllPathsExist(arguments, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
-
-            var fullProjectPaths = _arguments.Select(p =>
+            try
             {
-                var fullPath = Path.GetFullPath(p);
-                return Directory.Exists(fullPath) ?
-                    MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName :
-                    fullPath;
-            }).ToList();
-
-            var preAddProjectCount = slnFile.Projects.Count;
-
-            foreach (var fullProjectPath in fullProjectPaths)
+                PathUtility.EnsureAllPathsExist(_projects, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
+                IEnumerable<string> fullProjectPaths = _projects.Select(project =>
+                {
+                    var fullPath = Path.GetFullPath(project);
+                    return Directory.Exists(fullPath) ? MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName : fullPath;
+                });
+                AddProjectsToSolutionAsync(solutionFileFullPath, fullProjectPaths, CancellationToken.None).Wait();
+                return 0;
+            }
+            catch (GracefulException)
             {
-                // Identify the intended solution folders
-                var solutionFolders = DetermineSolutionFolder(slnFile, fullProjectPath);
-
-                slnFile.AddProject(fullProjectPath, solutionFolders);
+                throw;
             }
-
-            if (slnFile.Projects.Count > preAddProjectCount)
+            catch (Exception ex) when (ex is not GracefulException)
             {
-                slnFile.Write();
+                {
+                    if (ex is SolutionException || ex.InnerException is SolutionException)
+                    {
+                        throw new GracefulException(CommonLocalizableStrings.InvalidSolutionFormatString, solutionFileFullPath, ex.Message);
+                    }
+                    throw new GracefulException(ex.Message, ex);
+                }
             }
-
-            return 0;
         }
 
-        private static IList<string> GetSolutionFoldersFromProjectPath(string projectFilePath)
+        private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, IEnumerable<string> projectPaths, CancellationToken cancellationToken)
         {
-            var solutionFolders = new List<string>();
-
-            if (!IsPathInTreeRootedAtSolutionDirectory(projectFilePath))
-                return solutionFolders;
-
-            var currentDirString = $".{Path.DirectorySeparatorChar}";
-            if (projectFilePath.StartsWith(currentDirString))
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
```

Comment:
Is this a new runtime concept? I don't see it in the SDK currently or in this PR, and a quick bing (google) search didn't reveal anything interesting. If so, should we put this behind #ifdefs?

Reply:
Which part are you refering to here?

Reply:
Oh yes. It is a new serializer for sln(x) solution files that was recently released as open source and is being used by VS and MSBuild. 
https://github.com/microsoft/vs-solutionpersistence

Reply:
The goal long term is to replace `SlnFile` with this